src/r1cs/mod.rs

```diff
@@ -77,24 +70,31 @@ pub struct RelaxedR1CSInstance<E: Engine> {
 
 pub type CommitmentKeyHint<E> = dyn Fn(&R1CSShape<E>) -> usize;
 
-impl<E: Engine> R1CS<E> {
-  /// Generates public parameters for a Rank-1 Constraint System (R1CS).
-  ///
-  /// This function takes into consideration the shape of the R1CS matrices and a hint function
-  /// for the number of generators. It returns a `CommitmentKey`.
-  ///
-  /// # Arguments
-  ///
-  /// * `S`: The shape of the R1CS matrices.
-  /// * `ck_floor`: A function that provides a floor for the number of generators. A good function
-  ///   to provide is the ck_floor field defined in the trait `RelaxedR1CSSNARKTrait`.
-  ///
-  pub fn commitment_key(S: &R1CSShape<E>, ck_floor: &CommitmentKeyHint<E>) -> CommitmentKey<E> {
-    let num_cons = S.num_cons;
-    let num_vars = S.num_vars;
-    let ck_hint = ck_floor(S);
-    E::CE::setup(b"ck", max(max(num_cons, num_vars), ck_hint))
-  }
+/// Generates public parameters for a Rank-1 Constraint System (R1CS).
+///
+/// This function takes into consideration the shape of the R1CS matrices and a hint function
+/// for the number of generators. It returns a `CommitmentKey`.
+///
+/// # Arguments
+///
+/// * `S`: The shape of the R1CS matrices.
+/// * `ck_floor`: A function that provides a floor for the number of generators. A good function
+///   to provide is the ck_floor field defined in the trait `RelaxedR1CSSNARKTrait`.
+///
+pub fn commitment_key<E: Engine>(
+  S: &R1CSShape<E>,
+  ck_floor: &CommitmentKeyHint<E>,
+) -> CommitmentKey<E> {
+  let size = commitment_key_size(S, ck_floor);
+  E::CE::setup(b"ck", size)
+}
+
+/// Computes the number of generators required for the commitment key corresponding to shape `S`.
+fn commitment_key_size<E: Engine>(S: &R1CSShape<E>, ck_floor: &CommitmentKeyHint<E>) -> usize {
```

Comment:
Does it make sense to pass a vector/slice of R1CSShape objects to `commitment_key` method and have it pick the right size? This will avoid having two methods and logic above this layer?

Reply:
You can check the use of `commitment_key_size` in #283 in the `src/supernova/mod.rs`, in `compute_primary_ck`. This is left for a further PR so as to make those further changes in supernova-specific files, according to the general thrust of your [comments](https://github.com/microsoft/Nova/pull/283#issuecomment-1875777020) in #283.

Reply:
Thanks for the clarification! I think we could still have `commitment_key` to be a method inside R1CSShape rather than having two separate methods. For now, it takes (S: &R1CSShape<E>, ck_floor: &CommitmentKeyHint<E>) as arguments and returns a ck. In the future, it can take a collection of R1CSShape and hints and returns a ck. 

In other words, I don't see what is gained from having two methods, one called `commitment_key` and another called `commitment_key_size`. Why not inline the contents of `commitment_key_size` inside `commitment_key` method and then move `commitment_key` to be a method under `R1CSShape` type rather than a stand-alone method.